### ISSUE TITLE
Question on pre-train llama-66m on C4

### ISSUE THREAD

**Issue by zqOuO** [Wongboo/BlockwiseLearningRate#1]
Hello author, I am trying to pre-train LLaMA-66M on the C4 dataset. I want to make sure my replication is correct: 

Regarding the learning rate, I tried several base learning rates, such as 1e-3. First, I used linear warmup for 1k steps to ensure that the learning rate of all modules is 5e-4 at that point, and then multiplied each module by the learning rate ratios given in the paper, for example, the embedding becomes 5e-3 before cosine decay to 5%. For AdamW, I used betas of 0.9 and 0.95, with a weight decay of 0.1. The total number of training iterations was 10k. 

r(Emb) = 10, r(QK) = 8, r(FFN) = 6, r(VO) = 4.
learning rate(Emb) at iteration 1k= 10 * lr_base, learning rate(QK) = 8* lr_base, ....

Is my implementation correct?

{
    "architectures": ["LLaMAForCausalLM"],
    "bos_token_id": 0,
    "eos_token_id": 1,
    "hidden_act": "silu",
    "hidden_size": 512,
    "intermediate_size": 2048,
    "initializer_range": 0.02,
    "max_sequence_length": 1024,
    "model_type": "llama",
    "num_attention_heads": 8,
    "num_hidden_layers": 8,
    "pad_token_id": -1,
    "rms_norm_eps": 1e-06,
    "transformers_version": "4.28.1",
    "use_cache": true,
    "vocab_size": 32000
}

optimizer:configure_optimizer:409 - group 0: name=emb len(params)=1 weight_decay=0.1 warmup_lr=0.000500 target_ratio=10.00
optimizer:configure_optimizer:409 - group 1: name=qk len(params)=16 weight_decay=0.1 warmup_lr=0.000500 target_ratio=8.00
optimizer:configure_optimizer:409 - group 2: name=vo len(params)=16 weight_decay=0.1 warmup_lr=0.000500 target_ratio=4.00
optimizer:configure_optimizer:409 - group 3: name=mlp len(params)=24 weight_decay=0.1 warmup_lr=0.000500 target_ratio=6.00
optimizer:configure_optimizer:409 - group 4: name=other len(params)=18 weight_decay=0.1 warmup_lr=0.000500 target_ratio=1.00

Are there any other experimental details?

**Comment by wang-jinbo** [comment]
Seems to be right. However, I did not know what `target_ratio` is explicitly in your code.

**Comment by zqOuO** [comment]
Thank you for your reply! 
In my code, `target_ratio` refers to the learning rate ratio mentioned in your paper (r(Emb) = 10, r(QK) = 8, r(FFN) = 6, r(VO) = 4). 
With a base learning rate of 5e-4, after the linear warm-up ends and the actual learning rate of each module becomes 5e-4, the actual learning rate of all modules will be multiplied by the corresponding `target_ratio`. For example, the learning rate for embedding will become 5e-3. However, doing this causes the loss to spike for a period of time after the warm-up and sometimes fails to recover. 
Is my implementation correct?

**Comment by wang-jinbo** [comment]
Generally speaking, differences in training settings can lead to significant differences.
First, in the case of [Wen at al.](https://arxiv.org/pdf/2509.02046?), they use a 130M model with 8e-3 learning rate in Figure 1 without observing spikes. 66M in your case can likely allow larger lr than 8e-3. However, this ultimately depends on the precise training configuration.
Second, our paper presents a general principle, ratios are first derived from small-scale experiments and then transferred to larger experiments under the same setting. In principle, the ratio should be adjusted across different training setups. In practice, we retune the ratio when moving to a real world large-scale experiments setting.

**Comment by zqOuO** [comment]
Thank you for your answer！
